### PR TITLE
Add list of containers to docker volume ls command

### PR DESCRIPTION
**Pull request by brahmaroutu**
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

This PR is to list containers that are using the volumes. 
ping @cpuguy83.

### CODE REVIEW

**Comment by cpuguy83** [comment]
I don't think this is desirable.
It will make the `volume ls` command slow and potentially extremely wide.
Getting a container list on `inspect`, though, seems ok -- though looping through the container list doesn't seem like the best approach to get that info.

**Comment by duglin** [comment]
@cpuguy83 what's the recommended way to determine the list of volumes I can safely delete because they're not being used?  Feels like asking people to do an inspect on each volume might be tedious. I agree that the output might be wide - we could consolidate it down to an "inUse" column (with 'yes/no' values), but leave the REST API with the full list.  Just brainstorming...

**Comment by cpuguy83** [comment]
@duglin `--filter dangling=true`

**Comment by duglin** [comment]
cool - didn't know it had a filter flag!

**Comment by brahmaroutu** [comment]
@cpuguy83 there is  use case with customers who would like to know the volume association to container. I can move this to inspect if you think that is the right place to go.

**Comment by brahmaroutu** [comment]
BTW, IMO most typically each volume is linked to one or two containers and the 'ls' command does not seem to look wide.

**Comment by calavera** [comment]
I think this information should be ported to `docker ps` instead of adding it to `docker volume ls`. It would be easier to manage and inspect, since we already have the volume information mounted into each container.

You could search containers by volumes with:

```
$ docker ps --filter volume=foo
```

And get a pretty output with:

```
$ docker ps --format "{{table .ID\t.Mounts}}`
```

File: integration-cli/docker_cli_volume_test.go [hunk 1]
```diff
@@ -87,23 +131,23 @@ func (s *DockerSuite) TestVolumeCliLsFilterDangling(c *check.C) {
 	out, _ := dockerCmd(c, "volume", "ls")
 
 	// No filter, all volumes should show
-	c.Assert(out, checker.Contains, "testnotinuse1\n", check.Commentf("expected volume 'testnotinuse1' in output"))
```
Comment (tiborvass) [inline integration-cli/docker_cli_volume_test.go]: why is this change needed now?

Comment (brahmaroutu) [inline integration-cli/docker_cli_volume_test.go]: @tiborvass 'volume ls' will have add column.
vagrant@ubuntu-14:/go/src/github.com/docker/docker$ docker volume ls
DRIVER              VOLUME NAME                                                        CONTAINERS
local               ecfc0e38f94ea46cac7cb2277857d740f80a0e79a095d97087c645adb714ec37   []
local               abc                                                                [/c2 /c1]

Comment (tiborvass) [inline integration-cli/docker_cli_volume_test.go]: oh right

Comment (cpuguy83) [inline integration-cli/docker_cli_volume_test.go]: Yeah, I'll say again I don't think this is a good idea.

Comment (brahmaroutu) [inline integration-cli/docker_cli_volume_test.go]: @cpuguy83 I can implement @calavera suggestion but I think it is needed to see container associations from the volume commands to make processing simpler. I agree with your point on this being slow and all, but I think maintaining such list and syncing would be major change compared to building the list dynamically. I think at the moment either we extend 'ls' or 'inspect' to help this use case.

Comment (cpuguy83) [inline integration-cli/docker_cli_volume_test.go]: To be clear, I'm +1 for `inspect`, -1 for `ls`.

**Comment by brahmaroutu** [comment]
@cpuguy83 Please review my changes, I moved the container list into inspect.

File: daemon/inspect.go [hunk 2]
```diff
@@ -238,3 +240,15 @@ func (daemon *Daemon) getDefaultNetworkSettings(networks map[string]*network.End
 	}
 	return settings
 }
+
+func getContainersWithVolume(volume string, containers []*container.Container) []string {
+	var containerNames []string
+	for _, container := range containers {
```
Comment (calavera) [inline daemon/inspect.go]: This double iteration is going to make the command very slow when you have lots and lots of containers.

**Comment by brahmaroutu** [comment]
@calavera I agree with you. But we need a data model change to capture this association either through some notification model into container framework to signal attach/detach of volumes, etc and maintain that list as part of volume or seperately. I can work on such PR to optimize this. For now this is the best approach I can think of. Please let me know if you have any ideas that I can explore.

**Comment by cpuguy83** [comment]
@brahmaroutu In #16534, instead of using a counter for the volume references, I actually store the ID's -- this approach could be used here (and shorten up that PR a little :) )

**Comment by brahmaroutu** [comment]
@cpuguy83 Sorry, can you please elaborate your comment. I am iterating containers and not volumes. Do you want me to check the ref count before iterating to avoid unnecessary processing?

**Comment by cpuguy83** [comment]
@brahmaroutu You can pull in some of the changes from that PR which stores the container references for each volume instead of just a count.

**Comment by thaJeztah** [comment]
@brahmaroutu do you need more information, or are the comments clear? (just a friendly check)

**Comment by brahmaroutu** [comment]
@thaJeztah I will work on it, I will pull what is required from the other PR to implement container refs into volumes today.

**Comment by thaJeztah** [comment]
Looks like this was updated since the last comments?

ping @calavera @cpuguy83 can you have another look?

**Comment by calavera** [comment]
Sorry, but I still don't like this change :pensive:

We already have the information about volumes attached to containers in the container payload. There is no need to duplicate it. That information would be more useful for everyone if we exposed it in `docker ps`, like I suggested in the beginning.

> But we need a data model change to capture this association either through some notification model into container framework to signal attach/detach of volumes, etc and maintain that list as part of volume or seperately.

As I said, that information is stored with each container.

> For now this is the best approach I can think of. Please let me know if you have any ideas that I can explore.

This change is very inconsistent. Why would people use `docker ps` to get all kind of information about containers but `docker volume ls` to get information about volumes attached to containers? `docker ps` should be the source of truth to any information related to containers, and volumes attached to containers is part of that information. `docker volume ls` is the command to list volumes in the system, attached to containers or not, it's not its responsibility to know which volumes are attached to containers.

This change does NOT LGTM.

**Comment by thaJeztah** [comment]
@brahmaroutu I just discussed this PR with @calavera and we're looking for a use-case that cannot be satisfied with the approach @calavera mentioned here; https://github.com/docker/docker/pull/18312#issuecomment-160718282. 

I'll close this PR for now, but interested to hear if there's reasons that we're overlooking

**Comment by duglin** [comment]
For those interested, some IRC chat on this is here: https://botbot.me/freenode/docker-maintainers/2015-12-29/?msg=56951384&page=1

IMO managing volumes via "docker ps" is backwards and non-UX-intuitive. See the chat so I don't have to repeat it here.

**Comment by calavera** [comment]
I think we should reconsider this feature.

I realized this morning that I've been using a similar approach in #18888. Volumes and networks are the source of truth for their events, turning containers into additional data. Everything might be more consistent in the end if volumes know about containers attached to them too.

I'm reopening it.

**Comment by thaJeztah** [comment]
@calavera design LGTY? Can we move this to code review?

Also, I think this is now on `docker volume inspect`, no longer on `docker volume ls` or am I mistaken (if so the title of this PR needs to be edited)

**Comment by calavera** [comment]
We merged a change yesterday that allows you to know which volumes have containers connected and which one do not with one simple command #19671.

`docker volume ls --filter dangling=false` gives you exactly the list of volumes attached to containers.

I don't think we need this anymore.

**Comment by icecrime** [comment]
@duglin Do you agree with the comment above?

**Comment by duglin** [comment]
While that `dangling=true` is a good feature, its still very limiting. For example, I'm not sure how I would find out simple things like:
- how many volumes are used by more than one container?
- which containers are using a particular volume?
w/o looping over all my containers and doing `docker inspect` each one, which as I said above, feels backwards to me when volumes are supposed to be first class citizens.

**Comment by thaJeztah** [comment]
> how many volumes are used by more than one container?

I don't see that as an important feature, but I can relate to the second one (which containers are using a volume)

**Comment by cpuguy83** [comment]
I think it makes sense to have the list of containers that use a volume in the `volume inspect` output.
We're already tracking this internally.

**Comment by thaJeztah** [comment]
I think it makes sense to have this in `docker volume inspect`; moving this to code review

**Comment by thaJeztah** [comment]
ping @brahmaroutu can you rebase/update this PR?

**Comment by icecrime** [comment]
I don't think this PR is getting any attention anymore. I'll close it, but feel free to open a new one or ping us to reopen this one!
